src/pages/DealPreparation.tsx

```diff
@@ -0,0 +1,160 @@
+import { ProviderSelector } from "../components/ProviderSelector";
+import { Input as DealProposal, type Validated } from "../lib/dealProposal";
+import { uploadFile } from "../lib/fileUpload";
+import { callProposeDeal, callPublishDeal } from "../lib/jsonRpc";
+import { queryPeerId } from "../lib/requestResponse";
+
+export function DealPreparation({
+  account,
+}: {
+  account: InjectedAccountWithMeta;
+}) {
+  const [dealProposal, setDealProposal] = useState(
+    DealProposal.default().copyUpdate("client", account.address),
+  );
+
+  const [dealFile, setDealFile] = useState<File | null>(null);
+  const [providers, setProviders] = useState(new Map<string, object>());
+  const [selectedProviders, selectProviders] = useState<Set<string>>(new Set());
+  const [loading, setLoading] = useState(false);
+
+  const updateProviderSelection = useCallback((newProvider: string) => {
+    selectProviders((oldState) => {
+      const newSet = new Set(oldState);
+      if (newSet.has(newProvider)) {
+        newSet.delete(newProvider);
+      } else {
+        newSet.add(newProvider);
+      }
+      return newSet;
+    });
+  }, []);
+
+  const ctx = useCtx();
+
+  const performDeal = async (p: string, validDealProposal: Validated) => {
+    // We can make a function out of this and use Promise.all for more efficiency
+    const provider = providers.get(p);
+    if (!provider) {
+      // NOTE: this shouldn't really happen unless state management is wrong
+      throw new Error("Selected provider does not exist");
+    }
+
+    // TODO: type this
+    const providerPeerId = provider.peerId;
+    // TODO: catch the error?
+    const peerIdMultiaddress = await queryPeerId(providerPeerId);
+    if (!peerIdMultiaddress) {
+      toast.error(`Failed to find multiaddress for PeerId: ${providerPeerId}`);
+      return;
+    }
+
+    const nodeAddress = peerIdMultiaddress.nodeAddress();
+    const address = {
+      ip: nodeAddress.address,
+      port: nodeAddress.port,
+    };
+
+    const proposeDealResponse = await callProposeDeal(validDealProposal.toRpc(p), {
+      ...address,
+      port: 8000, // We can't hardcode this but we can't do anything about it *right now*
+    });
+
+    if (proposeDealResponse instanceof Error) {
+      toast.error(proposeDealResponse.message);
+      return;
+    }
+
+    if (!dealFile) {
+      // NOTE: This should never happen unless the "Continue" button is wrong
+      throw new Error("No file was selected");
+    }
+    await uploadFile(
+      dealFile,
+      proposeDealResponse,
+      { ...address, port: 8001 }, // We can't hardcode this but we can't do anything about it *right now*
+    );
+
+    const signedRpc = await validDealProposal.toSignedRpc(p, ctx.registry, account);
+    const publishDealResponse = await callPublishDeal(
+      signedRpc,
+      { ...address, port: 8000 }, // We can't hardcode this but we can't do anything about it *right now*
```

Comment:
```suggestion
      { ...address, port: DEFAULT_RPC_PORT }, // We can't hardcode this but we can't do anything about it *right now*
```

for some reason sometimes it's 8000, sometimes it's 8001?

Reply:
Yes, one is the HTTP upload server and the other is the JSON RPC